### ISSUE TITLE
Feature request: exclude with right mouse click menu

### ISSUE THREAD

**Issue by flip111** [shundhammer/qdirstat#69]
Could an option be implemented on the context menu that shows up under the right mouse button click? I like to exclude directories and files that way. Rescanning the disk is not needed in this case. Then i like to export this list to a file to use it as the tar `--exclude-ignore=` option. This is a great way to exclude big files and backup the rest.

By the way there seems to be a small bug in the exclude list at the moment. When the list is empty and i type something in "exclude directories with pattern" and then click on the Plus symbol a new item without text shows up. Then i have to re-type the text again to make it show up in the item. Why is the text not added in first instance? (Qdirstat 1.4)

I think a big textarea would be a bit more handy than using all the GUI buttons

Q: Is Wildcards compatible with the format `--exclude-ignore` expects ? (this should be similar to `.gitignore`)

Q: When adding a new rule (with Plus) could Pattern syntax be set to whatever it was set to when editing the last rule? Right now it seems to default on regular expression

Q: (a bit offtopic but didn't want to start a new issue): Can the stop reading button be placed on the icon bar?

Q: would it be possible to add an option to match to the relative directory? Or possibly read `.ignore` or `.gitignore` files?

Q: could the ignore list be auto-sorted? (or a textarea so i can sort it myself in a text editor)

Q: is the `/**/` wildcard supported for matching directories with arbitrary depth?

I noticed the list is stored in ~/.config/QDirStat-exclude.conf ... that's great ! that means i can hack my way around it and prepare the list before i start QDirStat :D

**Comment by flip111** [comment]
In the meantime i put together this script .. which is good enough to get me going. But i need to run it while QDirStat is closed because it doesn't reload and also it writes out the old conf on save. Maybe it's useful for anyone else...

```php
<?php

// config
$sort_paths = true;
$qdirstat_exclude_conf = '/home/flip111/.config/QDirStat/QDirStat-exclude.conf';
$rel_path = '/home/flip111'; // inspecting my home directory
$add_paths =
  [ '/Documents/websites' // some extra paths, leading slash will be a fixed path
  , 'Bla*Bla' // containing a star will use WildCards
  ];

// main script
$rel_path = rtrim($rel_path, '/');
$ini = parse_ini_file($qdirstat_exclude_conf, true);

function makeIniFile($ini) {
  $text_sections = [];

  foreach ($ini as $name => $section) {
    $out = sprintf("[%s]\n", $name);
    foreach ($section as $k => $v) {
      $out .= sprintf("%s=%s\n", $k, $v === '1' ? 'true' : $v);
    }
    $text_sections[] = $out;
  }

  return implode("\n", $text_sections);
}

function makeIgnoreFormat($ini, $rel_path) {
  $rel_path_len = strlen($rel_path);
  $paths = [];

  foreach ($ini as $section) {
    if (substr($section['Pattern'], 0, $rel_path_len) === $rel_path) {
      $v = substr($section['Pattern'], $rel_path_len);
    } else {
      $v = $section['Pattern'];
    }
    $paths[] = $v;
  }

  return $paths;
}

// $rel_path should not have a trailing slash !
function toIniFormat($paths, $rel_path) {
  $i = 1;
  $sections = [];

  foreach ($paths as $path) {
    if (substr($path, 0, 1) === '/') {
      $pattern = $rel_path . $path;
      $useFullPath = 'true';
    } else {
      $pattern = $path;
      $useFullPath = 'false';
    }

    $sections[sprintf('ExcludeRule_%02d', $i++)] =
      [ 'CaseSensitive' => 'true'
      , 'Pattern' => $pattern
      , 'Syntax' => strpos($path, '*') !== false ? 'Wildcard' : 'FixedString'
      , 'UseFullPath' => $useFullPath
      ];
  }

  return $sections;
}

$paths = makeIgnoreFormat($ini, $rel_path);
$paths = array_merge($paths, $add_paths);
$paths = array_unique($paths); // prevents double entries when using $add_paths
if ($sort_paths) {
  sort($paths); // optional sorting
}

file_put_contents($qdirstat_exclude_conf, makeIniFile(toIniFormat($paths, $rel_path)));
```

**Comment by shundhammer** [comment]
> Q: When adding a new rule (with Plus) could Pattern syntax be set to whatever it was set to when editing the last rule? Right now it seems to default on regular expression

Probably yes. Come to think of it, "regexp" is not a good default anyway since we are dealing with filenames here where "wildcard" would fit much better. But you are probably right, when the user selected something different recently that should remain selected. I'll look into that.

**Comment by shundhammer** [comment]
> Q: (a bit offtopic but didn't want to start a new issue): Can the stop reading button be placed on the icon bar?

It _could_ of course since those things are all QActions that can be placed anywhere in any toolbar, pulldown menu, popup menu.

But that toolbar is already getting cluttered; putting more and more stuff there would be counterproductive IMHO. I don't think many users need this one very often; once a directory tree is read for the first time, those directories are in the kernel's caches, and the next subsequent run is _very_ quick, even for an entire root filesystem. So, in most cases, you never get a chance to click there anyway because it's finished so quickly.

But if you find yourself needing it very often, I suggest you use its hotkey.

...which just made me realize that this one is not unique; it's Ctrl-C, but that's also "Edit -> Copy URL". Oops. Another small issue to fix.

**Comment by shundhammer** [comment]
> Q: would it be possible to add an option to match to the relative directory? 

Relative to what? To the directory you started QDirStat with? That might change for every program start; that's why this option is not there.

But depending on what exactly you want here, it might be possible with a clever regexp against the full path, or even with just a wildcard. The wildcards used here are not quite as limiting as the ones you use in a shell; you could write something like

    */foo/*/.git

which would correspond to a wildcard some advanced shells (e.g. zsh, IIRC also bash) offer:

    */foo/**/.git

meaning there might be any number of directories between "foo" and ".git".

> Or possibly read .ignore or .gitignore files?

No. Not only would that get complicated in a hurry, it might also open security concerns if somebody maliciously added stuff to such a .gitignore or .ignore file affecting files deeper down. You might never realize that you are not seeing certain things since you might never open that directory branch that is affected, not seeing an excluded directory there (and it would only exclude directories anyway like explained some answers before). You might think that you are only deleting useless stuff with a cleanup action without becoming aware that you just nuked your precious files there along with it.

Mind you, that danger also exists with exclude rules you write yourself, but the problem would become some dimensions greater if QDirStat were to pick up more of those exclude rules on the way while it reads directories.

While QDirStat would simply ignore those excluded directories, any cleanup action started as an external command would not. They would simply process those excluded directories as well, and if it's `/bin/rm` you are executing, this might prove desastrous.

**Comment by shundhammer** [comment]
> Q: is the /**/ wildcard supported for matching directories with arbitrary depth?

Already answered above... you'd write just `/*/` in this case (which might be a bit surprising if you only want to match one single directory, but that's how those QRegExp wildcards work).

**Comment by shundhammer** [comment]
> By the way there seems to be a small bug in the exclude list at the moment. When the list is empty and i type something in "exclude directories with pattern" and then click on the Plus symbol a new item without text shows up. Then i have to re-type the text again to make it show up in the item. Why is the text not added in first instance? (Qdirstat 1.4)

Yes, it is a bug, but the bug is really that this input field is enabled if there is no rule at all. You should not be able to enter anything if there is no current rule to edit.

I'll look into that.

**Comment by shundhammer** [comment]
> I think a big textarea would be a bit more handy than using all the GUI buttons

No. It's not only the exclude rule's text you edit, it's also its regexp mode:

- fixed string
- wildcard
- regexp

and what to compare against:

- full path
- directory name only

This works with the MIME Categories configuration only because there are only two possible combinations, case sensitive or case insensitive. Here we'd have six to cover all the possible combinations. That would be VERY confusing. Plus, as described above, the order matters, too, so having multiple large text areas would simply not work.

**Comment by flip111** [comment]
> so I'd restrict it to those simple cases

Yes this is fine, fullpath with fixed strings. User can edit the rule later. After this has been chosen menu option should change to something like "cancel exclude".

> But would you clutter a directory listing with possibly lots of excluded individual files

I hadn't thought about it before but now that you ask ... i would make a special entry (special with icon / background color or something) and group all the excluded files in there. It would work the same as a sub-directory but visually be different enough.

By the way .. what about hiding empty folders?

> Applying exclude rules has a performance penalty since every time an item is checked against them, all rules are applied in sequence

After you retrieved the path once you can run all the rules (so no extra syscalls). The syscall should be much more expensive. I don't know what can be done with cache locality and fast regex engines. Would need some benchmarks. For me personally i would be willing to wait quite a long time for QDirStat to have applied the rules on files too ...

> This is the background why this feature is there.

I can understand that big directories are more often present than large files ..

> you can only ask them to do that on a regular basis to avoid too much cruft to accumulate and thus fill up everybody's shared disk space

For my scenario i care mostly about small backups and not accumulating files in regular storage .. (they can be excluded after all).

> How generic could it be to cater every (or at least most) potential users' needs?

I would suggest following the unix philosophy here a little bit ... have 1 tool to do 1 job. Right now QDirStat is just barely (through that `.conf` files) giving me an option to do just that. Maybe a good question to ask is how well QDirStat can interface with tar and backup tools which are already out there so you don't need to implement all of that.

> Update: Fixed. It is now Ctrl-S.

I would expect `Ctrl-S` to open up a save dialog.

> That might change for every program start; that's why this option is not there.

Yeah makes sense.. This "request" only makes sense if QDirStat would load conf files form each directory (just like .gitignore) .. but i don't think that should be implemented .. So bad idea on my part ..

> You might think that you are only deleting useless stuff with a cleanup action

How would you think that if you didn't open the directory branch that contains the excluded directory (described in the sentence before). I'm just trying to grasp how you work with the program here ..

> any cleanup action started as an external command would not

I'm not familiar with the external command yet ....

> So the rules with the most likelihood of matching should come first, then the more exotic ones. This cannot just be auto-sorted.

There can be a different sorting internally than displayed to the user .. but ok the order starts to matter for when you want to build in the "ignore exclude rule" later. Maybe a sort button could be added though ..

> which might be a bit surprising if you only want to match one single directory

Is there anything that could be done to match a single directory? I think this is a very dangerous feature at the moment !!

----

I will close this now because the conf file was enough to get me going and questions have been answered. Thanks for the great program and support !

**Comment by shundhammer** [comment]
> > But would you clutter a directory listing with possibly lots of excluded individual files
>
> I  hadn't thought about it before but now that you ask ... 

No, I didn't ask. I explained why this would be a bad idea.

> i would make a special entry (special with icon / background color or something) and group all the excluded files in there. It would work the same as a sub-directory but visually be different enough.

Thus complicating the user interface with somebody that no user knows from any other application? Bad idea.

And it would mean that in most cases, the `<Files>` pseudo-directory that groups the files in a directory (as opposed to any subdirectories on that level) would need to get another tree level for those excluded files, or they would have to go to yet another pseudo directory in parallel to `<Files>`. 

No. Those files consume disk space. Showing disk space is what QDirStat is all about, so it shows how much disk space they use.

If the files are not useful to you, you might want to delete them. If they are, they are shown.

**Comment by shundhammer** [comment]
> > How generic could it be to cater every (or at least most) potential users' needs?
>
> I would suggest following the unix philosophy here a little bit ... have 1 tool to do 1 job. 

Exactly. QDirStat's job is to _to show where your disk space has gone and to help you to clean it up_ as it clearly says in the summary on its home page and in the "About" dialog.That's what it does; that's what it is meant for.

> Right now QDirStat is just barely (through that .conf files) giving me an option to do just that. 

If you managed to press it into service as a poor man's backup management tool, okay; users tend to get creative with what they do with software. I try to accomodate those users, too, if that doesn't get in the way of QDirStat's main job, but this is clearly not its main job.

> Maybe a good question to ask is how well QDirStat can interface with tar and backup tools which are already out there so you don't need to implement all of that.

So you want to use good old `tar`, fine. Others will probably want to use good old `rsync` since it can do quite some other tricks (including avoiding too much duplication if files remain unchanged). Still others will have other tricks up their sleeve what underlying tools to use.

Do you see a pattern here? 😃 

People are different. They have different preferences, different use cases, different views of the world. One's favourite tool is somebody else's old and outdated cruft. You just can't be everybody's darling, in particular not with off-topic uses like pressing a tool like QDirStat into service as a backup management GUI.

**Comment by shundhammer** [comment]
> > Update: Fixed. It is now Ctrl-S.
>
> I would expect Ctrl-S to open up a save dialog.

To save _what_? QDirStat is not an editor. It doesn't have semantics that come even close. There is not current document to save; what comes closes to the concept of a document in other applications is the directory tree, and this is nothing that can be saved.

**Comment by shundhammer** [comment]
> > You might think that you are only deleting useless stuff with a cleanup action
>
> How would you think that if you didn't open the directory branch that contains the excluded directory (described in the sentence before). I'm just trying to grasp how you work with the program here ..

If you just opened the `/home/kilroy/misc` directory branch, but you excluded all `foo` directories, you might think that you are just killing `/home/kilroy/misc` if you start the "delete directly" cleanup action from there (which will just execute the `rm -rf` command), but since you don't see `/home/kilroy/misc/aaa/foo` there anymore since it's excluded, you might not be aware that it exists since you don't see it.

In general, this applies to _all_ cleanup actions execute a command that works recursively in any way. Commands that are executed repeatedly by QDirStat (where the user checked `recursive` in the cleanup configuration) are fine, though, since they won't be called in excluded directories.

Come to think of it, it might be a good idea to explicitly ask the user once more before executing a cleanup action if a directory tree contains excluded directories.

**Comment by shundhammer** [comment]
> > which might be a bit surprising if you only want to match one single directory
>
> Is there anything that could be done to match a single directory? 

Just match the full path, of course.

**Comment by flip111** [comment]
> Do you see a pattern here? :smiley:

Yes `|`

> There is no current document to save

It's up to you.. i would reserve this for future use in case you do want to save something. People are used to CTRL + S from other applications. They don't know this as "stop current processing".

> but since you don't see /home/kilroy/misc/aaa/foo there anymore since it's excluded

Would the user have to click on the `aaa` folder to see `foo` there with a changed icon?

> Just match the full path, of course.

I meant in a pattern. When you want to match `foo/*/bar` where `*` should be one directory (i called it a "single directory", maybe that was not the right wording). As i understand you now `*` can expand to several directories like `/foo/qux/quix/bar`.

**Comment by shundhammer** [comment]
The way QRegExp's wildcards work I think you'd have to use a real regular expression for that (which sucks, I know, but that's just how QRegExp's wildcards work):

`.*/foo/[^/]+/bar`

(not tested, but I am pretty sure this should work)

Essentially, this matches "anything, then a slash, then "foo", then again a slash, then more than one character that isn't a slash, then a slash, then "bar"".

**Comment by shundhammer** [comment]
See also issue #92. That approach might also work for you.

**Comment by stdedos** [comment]
Could the basic case of "exclude this directory path" be implemented?

My usecase is I want to write exclude directory rules to my `/home` backup. If I keep removing the big directories (VM folders, repos, videos etc), then I will keep seeing the "biggest" offenders and decide if I need to exclude more or not (since conserving backup space / necessity is not the small directories, but the big ones).

It might be worth to do it so that, they can be toggled on or off; while still saving the full scan, if the user chooses to export the file.

The "exclude this directory (path)" feature already exists in TreeSize (https://www.jam-software.com/treesize_free/); unfortunately this is a Windows-only show

**Comment by shundhammer** [comment]
The approach of issue #92 should also work for this use case:

Create a cleanup action that does nothing, but that has a refresh policy of "assume deleted", and you can cut off branches of a directory tree very easily. You can even add a keyboard shortcut (by editing the config file).

See issue #92 (scroll all the way down) for more details.

**Comment by stdedos** [comment]
Apologies for the unnecessary comment! It turned out that there was a user before me here, trying to do the same thing (https://github.com/shundhammer/qdirstat/issues/92#issuecomment-472238733).

The topics (this and #92), are a little big; so, in the end I gave up on reading (since none of them were "fixed" by some commit, I thought that they were resolved in some "Can't/Won't Do" fashion).

Now all I need to figure out is "how the program works" and how to disable "the other" cleanup functions; I don't like things `rm`ing unless I press `Delete` and then `Ok` 😄 

Thank you for having such a diverse dirstat program!

**Comment by shundhammer** [comment]
The default cleanup actions are just the initial set that you get. You can simply remove each one of them with the "delete" button in the cleanup config dialog just like any user-created one.

If you change your mind and want to start over from scratch, you can simply remove that one config file; check out your `~/.config/QDirStat/` directory. All the files there human readable. `QDirStat-cleanup.conf` contains the cleanup actions; if you simply remove that one, the next program start will recreate it with the default ones.
